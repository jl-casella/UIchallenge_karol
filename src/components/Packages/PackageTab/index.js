import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

const PackageTabContainer = styled.div`
  display: inline-block;
  padding: 10px;
  font-weight: ${props => (props.active ? 'bold' : 'normal')};
`

const PackageTab = ({ number, active }) => (
  <PackageTabContainer active={active}>
    <a>Package {number}</a>
  </PackageTabContainer>
)

PackageTab.propTypes = {
  number: PropTypes.number.isRequired,
  active: PropTypes.bool
}

export default PackageTab
