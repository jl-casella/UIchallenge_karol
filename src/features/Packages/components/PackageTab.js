import Popup from 'reactjs-popup'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

const PackageTabContainer = styled.li`
  display: inline-flex;
  align-items: center;
  padding: 6px 8px;
  font-weight: ${props => (props.active ? 'bold' : 'normal')};
  background-color: ${props => (props.active ? 'lightgray' : 'white')};
  border-radius: 4px;
  margin: 4px 6px;
  font-size: 12px;
  border: 1px solid black;
  cursor: pointer;
`

const RemoveButton = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 14px;
  height: 14px;
  border-radius: 10px;
  border: 1px solid ${props => (props.disabled ? 'gray' : 'black')};
  color: ${props => (props.disabled ? 'gray' : 'black')};
  background-color: white;
  font-size: 8px;
  margin-left: 6px;
  font-weight: normal;

  :hover {
    background-color: ${props => (props.disabled ? 'white' : 'silver')};
  }
`

const PackageTab = ({
  number,
  active,
  setActivePackage,
  removeEnabled,
  removePackage
}) => {
  const onRemoveButtonClick = e => {
    if (removeEnabled) {
      removePackage()
    }

    e.stopPropagation()
  }

  return (
    <PackageTabContainer
      active={active}
      onClick={setActivePackage}
      removeEnabled={removeEnabled}
    >
      <a>Package {number}</a>
      <Popup
        trigger={
          <RemoveButton
            disabled={!removeEnabled}
            role="button"
            onClick={onRemoveButtonClick}
          >
            X
          </RemoveButton>
        }
        position="top center"
        on="hover"
        contentStyle={{
          width: 'auto',
          fontWeight: 'normal',
          textAlign: 'center'
        }}
      >
        <span>
          {removeEnabled
            ? 'Click to remove package'
            : 'Package must be emptied out before removing'}
        </span>
      </Popup>
    </PackageTabContainer>
  )
}

PackageTab.propTypes = {
  number: PropTypes.number.isRequired,
  active: PropTypes.bool,
  setActivePackage: PropTypes.func.isRequired,
  removePackage: PropTypes.func.isRequired,
  removeEnabled: PropTypes.bool.isRequired
}

export default PackageTab
