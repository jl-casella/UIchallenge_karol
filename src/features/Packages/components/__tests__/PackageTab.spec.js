import { noop } from 'lodash'
import React from 'react'
import { render, waitForElement, fireEvent } from '@testing-library/react'

import PackageTab from '../PackageTab'

describe('PackageTab component test', () => {
  it('Should render correctly when active', () => {
    const { container } = render(
      <PackageTab
        number={1}
        active={true}
        setActivePackage={noop}
        removePackage={noop}
        removeEnabled={true}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Should render correctly when inactive', () => {
    const { container } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={noop}
        removePackage={noop}
        removeEnabled={true}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Should render information tooltip when hovers over "X" button if removing disabled', async () => {
    const { getByRole, getByText } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={noop}
        removePackage={noop}
        removeEnabled={false}
      />
    )

    const removeButton = getByRole('button')
    fireEvent.mouseOver(removeButton)

    await waitForElement(() =>
      getByText('Package must be emptied out before removing')
    )
  })

  it('Should render information tooltip when hovers over "X" button if removing enabled', async () => {
    const { getByRole, getByText } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={noop}
        removePackage={noop}
        removeEnabled={true}
      />
    )

    const removeButton = getByRole('button')
    fireEvent.mouseOver(removeButton)

    await waitForElement(() => getByText('Click to remove package'))
  })

  it('Calls onRemove callback when clicking on "X" button if removing enabled', () => {
    const spy = jest.fn()
    const { getByRole } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={noop}
        removePackage={spy}
        removeEnabled={true}
      />
    )

    const removeButton = getByRole('button')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(removeButton)
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('Does not call onRemove callback when clicking on "X" button if removing disabled', () => {
    const spy = jest.fn()
    const { getByRole } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={noop}
        removePackage={spy}
        removeEnabled={false}
      />
    )

    const removeButton = getByRole('button')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(removeButton)
    expect(spy).toHaveBeenCalledTimes(0)
  })

  it('Calls onSetActive callback after clicking on package number', () => {
    const spy = jest.fn()
    const { getByText } = render(
      <PackageTab
        number={1}
        active={false}
        setActivePackage={spy}
        removePackage={noop}
        removeEnabled={false}
      />
    )

    const tabNameContainer = getByText('Package 1')
    expect(spy).toHaveBeenCalledTimes(0)

    fireEvent.click(tabNameContainer)
    expect(spy).toHaveBeenCalledTimes(1)
  })
})
