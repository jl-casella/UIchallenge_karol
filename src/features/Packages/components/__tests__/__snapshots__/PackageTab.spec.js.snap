// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageTab component test Should render correctly when active 1`] = `
<li
  class="PackageTab__PackageTabContainer-sc-1wzierc-0 buJVGS"
>
  <a>
    Package 
    1
  </a>
  <div
    class="PackageTab__RemoveButton-sc-1wzierc-1 jNsckS"
    role="button"
  >
    X
  </div>
</li>
`;

exports[`PackageTab component test Should render correctly when inactive 1`] = `
<li
  class="PackageTab__PackageTabContainer-sc-1wzierc-0 jolmYP"
>
  <a>
    Package 
    1
  </a>
  <div
    class="PackageTab__RemoveButton-sc-1wzierc-1 jNsckS"
    role="button"
  >
    X
  </div>
</li>
`;
